pkg/solana/logpoller/job.go

```diff
@@ -0,0 +1,149 @@
+package logpoller
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/gagliardetto/solana-go"
+	"github.com/gagliardetto/solana-go/rpc"
+)
+
+// Job is a function that should be run by the worker group. The context provided
+// allows the Job to cancel if the worker group is closed. All other life-cycle
+// management should be wrapped within the Job.
+type Job interface {
+	String() string
+	Run(context.Context) error
+}
+
+type retryableJob struct {
+	name  string
+	count uint8
+	when  time.Time
+	job   Job
+}
+
+func (j retryableJob) String() string {
+	return j.job.String()
+}
+
+func (j retryableJob) Run(ctx context.Context) error {
+	return j.job.Run(ctx)
+}
+
+type eventDetail struct {
+	blockNumber uint64
+	blockHash   solana.Hash
+	trxIdx      int
+	trxSig      solana.Signature
+}
+
+// processEventJob is a job that processes a single event. The parser should be a pure function
+// such that no network requests are made and no side effects are produced.
+type processEventJob struct {
+	parser ProgramEventProcessor
+	event  ProgramEvent
+}
+
+func (j *processEventJob) String() string {
+	return fmt.Sprintf("processEventJob for event: %s", j.event.FunctionName)
+}
+
+func (j *processEventJob) Run(_ context.Context) error {
+	return j.parser.Process(j.event)
+}
+
+// getTransactionsFromBlockJob is a job that fetches transaction signatures from a block and loads
+// the job queue with getTransactionLogsJobs for each transaction found in the block.
+type getTransactionsFromBlockJob struct {
+	slotNumber uint64
+	client     RPCClient
+	parser     ProgramEventProcessor
+	chJobs     chan Job
+}
+
+func (j *getTransactionsFromBlockJob) String() string {
+	return fmt.Sprintf("getTransactionsFromBlockJob for block: %d", j.slotNumber)
+}
+
+func (j *getTransactionsFromBlockJob) Run(ctx context.Context) error {
+	var excludeRewards bool
+
+	block, err := j.client.GetBlockWithOpts(
+		ctx,
+		j.slotNumber,
+		// You can specify more options here:
+		&rpc.GetBlockOpts{
+			Encoding:   solana.EncodingBase64,
+			Commitment: rpc.CommitmentFinalized,
+			// get the full transaction details
+			TransactionDetails: rpc.TransactionDetailsFull,
+			// exclude rewards
+			Rewards: &excludeRewards,
+		},
+	)
+	if err != nil {
+		return err
+	}
+
+	blockSigsOnly, err := j.client.GetBlockWithOpts(
+		ctx,
+		j.slotNumber,
+		// You can specify more options here:
```

Comment:
Is this comment referring to the possibility of us adding some more things to this code later, like a TODO? Or something else?